python/lsst/pipe/tasks/characterizeImage.py

```diff
@@ -0,0 +1,493 @@
+        """
+        self._frame = self._initialFrame # reset debug display frame
+        self.log.info("Processing %s" % (dataRef.dataId))
+
+        if doUnpersist:
+            if exposure is not None or background is not None:
+                raise pipeBase.TaskError("doUnpersist true; exposure and background must be None")
+            exposure = dataRef.get("postISRCCD", immediate=True)
+        elif exposure is None:
+            raise pipeBase.TaskError("doUnpersist false; exposure must be provided")
+
+        exposureIdInfo = dataRef.get("expIdInfo")
+
+        charRes = self.characterize(
+            exposure = exposure,
+            exposureIdInfo = exposureIdInfo,
+            background = background,
+        )
+
+        if self.config.doWrite:
+            dataRef.put(charRes.sourceCat, "icSrc")
+            if self.config.doWriteExposure:
+                dataRef.put(charRes.exposure, "icExp")
+                dataRef.put(charRes.background, "icExpBackground")
+
+        return charRes
+
+    @pipeBase.timeMethod
+    def characterize(self, exposure, exposureIdInfo, background=None):
+        """!Characterize a science image
+
+        Peforms the following operations:
+        - Iterate the following config.psfIterations times, or once if config.doMeasurePsf false:
+            - detect and measure sources and estimate PSF (see detectMeasureAndEstimatePsf for details)
+        - interpolate over cosmic rays
+        - perform final measurement
+
+        @param[in,out] exposure  exposure to characterize (an lsst.afw.image.ExposureF or similar).
+            The following changes are made:
+            - update or set psf
+            - set apCorrMap
+            - update detection and cosmic ray mask planes
+            - subtract background and interpolate over cosmic rays
+        @param[in] exposureIdInfo  ID info for exposure (an lsst.daf.butlerUtils.ExposureIdInfo)
+        @param[in] background  model of background model already subtracted from exposure
+            (an lsst.afw.math.BackgroundList). May be None if no background has been subtracted,
+            which is typical for image characterization.
+
+        @return pipe_base Struct containing these fields, all from the final iteration
+        of detectMeasureAndEstimatePsf:
+        - exposure: characterized exposure; image is repaired by interpolating over cosmic rays,
+            mask is updated accordingly, and the PSF model is set
+        - sourceCat: detected sources (an lsst.afw.table.SourceCatalog)
+        - background: model of background subtracted from exposure (an lsst.afw.math.BackgroundList)
+        - psfCellSet: spatial cells of PSF candidates (an lsst.afw.math.SpatialCellSet)
+        """
+        self._frame = self._initialFrame # reset debug display frame
+
+        if not self.config.doMeasurePsf and not exposure.hasPsf():
+            raise pipeBase.TaskError("exposure has no PSF model and config.doMeasurePsf false")
```

Comment:
This seems like something that the user would want a traceback for (so they can find where it's coming from), but using `TaskError` suppresses that.